docs/DesignDocs/AttestationEvidenceFormatSupport.md

```diff
@@ -0,0 +1,118 @@
+2. Call the new API to verify the evidence. The format id is not required because it is contained in the evidence buffer.
+
+Scenario 1 is for users who provide their own evidence structures (like #2999, user hopes to verify a raw sgx quote).
+
+Scenario 2 is for users who full rely in OESDK ecosystem. They can use oe_generate_evidence to generate an oe_evidence structure and verify it by oe_verify_evidence.
+
+In scenario 2, users should pass a `null` format id so that new `oe_get_evidence()` avoids guessing the provided evidence is a raw evidence or an oe_evidence, and the cases that the provided format id is not consistent with the id contained in the oe_evidence buffer.
+```
+Attester				        Verifier
+--------- 				        ----------
+1. SGX Quote 				        a. oe_verify_evidence(SGX_P256, evidence, ...)
+2. oe_get_report() 			        b. oe_verify_evidence(OE_REPORT_ID, evidence, ...)
+3. oe_get_evidence(SGX_P256) 	                c. oe_verify_evidence(null, evidence, ...)
+
+Evidence 1 should be verified by calling function a.
+Evidence 2 should be verified by calling function b.
+Evidence 3 should be verified by calling function c, not a.
+```
+In both scenarios, the plugin_verifier is located by format_id and then verifies the evidence. But the evidence in scenario 2 might contain a custom claim buffer which is added by `oe_get_evidence()`. The plugin_verifier needs to do an extra size check to tell the difference. For example, `oe_verify_evidence(SGX_P256, evidence, ...)` will locate the verifier to verify the `evidence`. The verifier needs to check if the evidence size equals to the raw sgx quote size, or `if (evidence_size == ((sgx_quote) evidence).size)`. If it is, then there is no extra claim buffer. The verifier's claim extraction function should take care of that.
+
+
+Specification
+-------------
+
+### New API
+An extra parameter `format_id` is added to current function oe_verify_evidence().
+
+```
+oe_result_t oe_verify_evidence(
+    const oe_uuid_t* format_id,
+    const uint8_t* evidence_buffer,
+    size_t evidence_buffer_size,
+    const uint8_t* endorsements_buffer,
+    size_t endorsements_buffer_size,
+    const oe_policy_t* policies,
+    size_t policies_size,
+    oe_claim_t** claims,
+    size_t* claims_length);
+```
+This API is used to verify both evidences whose format id is provided by verifier and who is generated by `oe_get_evidence()`. When actual format_id is provided, it decouples evidence format ID and raw evidence buffer. When format_id is `null`, the API acts the same as previous `oe_get_evidence()`. This new design will provide more flexibility.
+
+
+### New Format ID
+According to issue #2999 reported by attestation service, it has to support evidence generated by SGX quote generation and `oe_get_report()`. SGX quote format id is already defined as `OE_FORMAT_UUID_SGX_ECDSA_P256`. A new id needs to be defined for evidence generated by `oe_get_report()`.
+
```

Comment:
Let's keep the existing `OE_FORMAT_UUID_SGX_ECDSA_P256` for the evidence generated by `oe_get_evidence()`, and define two new ones `OE_FORMAT_UUID_SGX_ECDSA_P256_REPORT` and `OE_FORMAT_UUID_SGX_ECDSA_P256_QUOTE`, for `oe_get_report()` output and raw SGX quote respectively.

I'm updating the design doc Attestation_API_Proposal.md and implementation to capture those changes.

Reply:
Will we have 3 different verifiers for these 3 evidences? And in the future, do we have to assigned two IDs for one evidence? One for raw evidence, another for the one generated by `oe_get_evidence()`.

Reply:
I see the need of 3 format IDs. The unpacked `oe_evidence` is `oe_report`, which in turn is SGX quote prefixed with a header.

I've just created PR #3136 with the changes in Attestation_API_Proposal.md.

Reply:
Well my previous design set same format id for a raw evidence and oe_evidence. But I think your design makes more sense.
Here is a summary:
| Format ID                            | Evidence                                                   |
|--------------------------------------|------------------------------------------------------------|
| `OE_FORMAT_UUID_SGX_ECDSA_P256`        | SGX ECDSA-p256 evidence   generated by `oe_get_evidence()` |
| `OE_FORMAT_UUID_SGX_ECDSA_P256_QUOTE`  | raw SGX ECDSA quote                                        |
| `OE_FORMAT_UUID_SGX_ECDSA_P256_REPORT` | SGX ECDSA report generated by `oe_get_report()`            |

If we follow this way, in future, suppose we want to support new evidence formats, such as VBS, we will generate **a pair of ids**
| Format ID                | Evidence                                        |
|--------------------------|-------------------------------------------------|
| `OE_FORMAT_UUID_VBS`       | VBS evidence generated by   `oe_get_evidence()` |
| `OE_FORMAT_UUID_VBS_QUOTE` | raw VBS quote                                   |

Pros:
1. One evidence format strictly maps to one format id and verifier. It makes a cleaner design.
2. Claims extraction can be differentiated. For example, a SGX ECDSA-p256 evidence gets a `OE_CUSTOM_CLAIMS` claim entry, a raw SGX ECDSA quote gets a `OE_CLAIM_SGX_REPORT_DATA` claim entry. With different verifiers, we can call different extraction procedure for each evidence format.

Cons:
1. `oe_verify_evidence` select the plugin verifiers by `format_id` so we need to define two different verifiers for a raw evidence and the oe_evidence containing the raw evidence. We need to double the format id and verifiers even two verifiers do mostly the same thing.
2. Users might get confused about the pair of ids. We need to document it well.

Reply:
I've updated based on our discussion. Two new format ids are introduced: "OE_FORMAT_UUID_SGX_ECDSA_P256_QUOTE" and "OE_FORMAT_UUID_SGX_ECDSA_P256_REPORT".